### PR TITLE
Add support for custom certificates

### PR DESCRIPTION
**Pull request by voczi**
Attempt nr 2 from last PR. Ankidroid parts are already done so I'll follow this PR up with a couple others once it's merged.

### CODE REVIEW

File: rslib/src/backend/ankiweb.rs [hunk 1]
```diff
@@ -26,6 +26,7 @@ impl Backend {
         self.runtime_handle().block_on(async move {
             let out = self
                 .web_client()
+                .unwrap()
```
Comment (dae) [inline rslib/src/backend/ankiweb.rs:29]: This should be a `?` instead to bubble the error up instead of crashing.

File: rslib/src/backend/mod.rs [hunk 2]
```diff
@@ -137,10 +137,54 @@ impl Backend {
             .clone()
     }
 
-    fn web_client(&self) -> &Client {
-        // currently limited to http1, as nginx doesn't support http2 proxies
```
Comment (dae) [inline rslib/src/backend/mod.rs:141]: Please preserve the comment.

File: rslib/src/backend/mod.rs [hunk 3]
```diff
@@ -137,10 +137,54 @@ impl Backend {
...
+        let mut client_mutex = self.web_client.lock();
+
+        if cert_str.is_empty() {
+            if let Ok(ref mut web_client) = client_mutex {
+                let _ = (**web_client).insert(Client::builder().http1_only().build().unwrap());
+                return Ok(());
+            }
+        }
+
+        #[cfg(feature = "rustls-pemfile")]
```
Comment (dae) [inline rslib/src/backend/mod.rs:156]: Is this cfg check needed, given that the feature is always included?

File: rslib/src/backend/mod.rs [hunk 4]
```diff
@@ -137,10 +137,54 @@ impl Backend {
...
-        self.web_client
-            .get_or_init(|| Client::builder().http1_only().build().unwrap())
+    #[cfg(feature = "rustls")]
+    fn set_custom_certificate_inner(&self, cert_str: String) -> Result<()> {
+        use std::io::Cursor;
+        use std::io::Read;
+
+        use reqwest::Certificate;
+
+        let mut client_mutex = self.web_client.lock();
```
Comment (dae) [inline rslib/src/backend/mod.rs:147]: Please .unwrap() .lock() calls, as a poisoned mutex should result in shutdown.

File: rslib/src/backend/sync.rs [hunk 5]
```diff
@@ -239,7 +249,7 @@ impl Backend {
             (col.media()?, col.new_progress_handler())
         };
         let rt = self.runtime_handle();
-        let sync_fut = mgr.sync_media(progress, auth, self.web_client().clone(), server_usn);
+        let sync_fut = mgr.sync_media(progress, auth, self.web_client().unwrap(), server_usn);
```
Comment (dae) [inline rslib/src/backend/sync.rs:252]: Callers of web_client() should bubble the error up, instead of unwrapping.

**Review by dae — COMMENTED** [review]
Almost there

File: rslib/src/backend/mod.rs [hunk 6]
```diff
@@ -137,10 +137,46 @@ impl Backend {
...
+    }
+
+    fn web_client(&self) -> Client {
         // currently limited to http1, as nginx doesn't support http2 proxies
-        self.web_client
-            .get_or_init(|| Client::builder().http1_only().build().unwrap())
+        let mut web_client = self.web_client.lock().unwrap();
+
+        return web_client
+            .get_or_insert(Client::builder().http1_only().build().unwrap())
```
Comment (dae) [inline rslib/src/backend/mod.rs:178]: Please use get_or_insert_with() so that a new client is not created when one already exists.

**Comment by dae** [comment]
Thanks!

**Comment by voczi** [comment]
Not sure why but "voczi <dev@voczi.com" wasn't included in the commits. Weird. Doesn't matter, though, I'll push the remaining code to ankidroid repos later today.

**Pull request merged**